File: 09_hull_white_short_rate_model/python/hull_white_model.py
"""
hull_white_model.py

One-factor Hull-White short-rate model.

Model:
------
dr(t) = [theta(t) - a r(t)] dt + sigma dW(t)

where:

- a     = mean reversion speed
- sigma = short-rate volatility
- theta(t) is chosen to fit the initial term structure

This chapter focuses on:
- curve-consistent short-rate dynamics
- discount bond pricing
- simulation
- bond option pricing
"""

import numpy as np


class HullWhiteModel:
    """
    One-factor Hull-White model.

    Parameters
    ----------
    mean_reversion : float
        Mean reversion speed a.

    volatility : float
        Short-rate volatility sigma.

    curve : YieldCurve
        Initial discount curve.
    """

    def __init__(
        self,
        mean_reversion,
        volatility,
        curve
    ):

        self.a = mean_reversion
        self.sigma = volatility
        self.curve = curve

    def B(self, t, T):
        """
        Hull-White B(t,T) function.

        B(t,T) measures sensitivity of the zero-coupon bond price
        to the short rate r(t).

        Formula:

            B(t,T) = (1 - exp(-a(T-t))) / a

        Parameters
        ----------
        t : float
            Current time.

        T : float
            Bond maturity.

        Returns
        -------
        float
            B(t,T).
        """

        if T < t:
            raise ValueError("Bond maturity T must be greater than or equal to t.")

        if abs(self.a) < 1e-12:
            return T - t

        return (
            1.0
            -
            np.exp(
                -self.a * (T - t)
            )
        ) / self.a

    def instantaneous_forward_rate(self, t, bump=1e-2):
        """
        Approximate instantaneous forward rate from the initial curve.

        The instantaneous forward rate is:

            f(0,t) = - d log P(0,t) / dt

        We compute it numerically using central differences.

        Parameters
        ----------
        t : float
            Time.

        bump : float
            Finite difference bump.

        Returns
        -------
        float
            Instantaneous forward rate.
        """

        t1 = max(t - bump, 1e-6)
        t2 = t + bump

        p1 = self.curve.discount_factor(t1)
        p2 = self.curve.discount_factor(t2)

        return -(
            np.log(p2)
            -
            np.log(p1)
        ) / (t2 - t1)

    def instantaneous_forward_derivative(self, t, bump=5e-2):
        """
        Numerical derivative of the instantaneous forward rate.

        This is used in the theoretical theta(t) expression.

        Parameters
        ----------
        t : float
            Time.

        bump : float
            Finite difference bump.

        Returns
        -------
        float
            df(0,t) / dt.
        """

        t1 = max(t - bump, 1e-6)
        t2 = t + bump

        f1 = self.instantaneous_forward_rate(t1)
        f2 = self.instantaneous_forward_rate(t2)

        return (f2 - f1) / (t2 - t1)

    def theta(self, t):
        """
        Hull-White drift function theta(t).

        For the model:

            dr = [theta(t) - a r] dt + sigma dW

        theta(t) is chosen so that the model fits the initial discount curve.

        A commonly used expression is:

            theta(t)
            =
            df(0,t)/dt
            + a f(0,t)
            + sigma^2/(2a) * (1 - exp(-2at))

        Returns
        -------
        float
            Drift adjustment theta(t).
        """

        f = self.instantaneous_forward_rate(t)

        dfdt = self.instantaneous_forward_derivative(t)

        if abs(self.a) < 1e-12:
            convexity = self.sigma * self.sigma * t
        else:
            convexity = (
                self.sigma
                *
                self.sigma
                /
                (2.0 * self.a)
                *
                (
                    1.0
                    -
                    np.exp(
                        -2.0 * self.a * t
                    )
                )
            )

        return (
            dfdt
            +
            self.a * f
            +
            convexity
        )

    def short_rate_drift(self, t, r):
        """
        Drift of the short rate process.

        Parameters
        ----------
        t : float
            Time.

        r : float
            Current short rate.

        Returns
        -------
        float
            Drift term.
        """

        return self.theta(t) - self.a * r

    def evolve_short_rate(
        self,
        t,
        r,
        dt,
        z
    ):
        """
        One Euler step for the Hull-White short-rate process.

        Parameters
        ----------
        t : float
            Current time.

        r : float
            Current short rate.

        dt : float
            Time step.

        z : float
            Standard normal random variable.

        Returns
        -------
        float
            Next short rate.
        """

        drift = self.short_rate_drift(
            t,
            r
        )

        return (
            r
            +
            drift * dt
            +
            self.sigma
            *
            np.sqrt(dt)
            *
            z
        )

File: 11_bermudan_swaption_tree_pricing/cpp/HullWhiteModel.cpp
#include "HullWhiteModel.hpp"

#include <cmath>
#include <stdexcept>

/*
    Constructor
*/

HullWhiteModel::HullWhiteModel(
    double meanReversion,
    double volatility,
    YieldCurve* curve_
){
    a = meanReversion;
    sigma = volatility;
    curve = curve_;
}

/*
    Hull-White B(t,T):

        B(t,T) = (1 - exp(-a(T-t))) / a

    This measures bond price sensitivity to the short rate.
*/

double HullWhiteModel::B(
    double t,
    double T
) const {

    if(T < t){
        throw std::runtime_error(
            "T must be greater than or equal to t."
        );
    }

    if(std::abs(a) < 1e-12){
        return T - t;
    }

    return
        (
            1.0
            -
            std::exp(
                -a * (T - t)
            )
        )
        /
        a;
}

/*
    Approximate instantaneous forward rate:

        f(0,t) = - d log P(0,t) / dt

    We use a practical bump because the curve is interpolated from
    discrete market pillars. Very small bumps may create numerical noise.
*/

double HullWhiteModel::instantaneousForwardRate(
    double t,
    double bump
) const {

    double t1 =
        std::max(
            t - bump,
            1e-6
        );

    double t2 =
        t + bump;

    double p1 =
        curve->discount_factor(t1);

    double p2 =
        curve->discount_factor(t2);

    return
        -(
            std::log(p2)
            -
            std::log(p1)
        )
        /
        (t2 - t1);
}

/*
    Numerical derivative of instantaneous forward rate.

        df(0,t) / dt
*/

double HullWhiteModel::instantaneousForwardDerivative(
    double t,
    double bump
) const {

    double t1 =
        std::max(
            t - bump,
            1e-6
        );

    double t2 =
        t + bump;

    double f1 =
        instantaneousForwardRate(t1);

    double f2 =
        instantaneousForwardRate(t2);

    return
        (f2 - f1)
        /
        (t2 - t1);
}

/*
    Hull-White theta(t):

        theta(t)
        =
        df(0,t)/dt
        + a f(0,t)
        + sigma^2/(2a) * (1 - exp(-2at))

    This drift term makes the model fit the initial yield curve.
*/

double HullWhiteModel::theta(
    double t
) const {

    double f =
        instantaneousForwardRate(t);

    double dfdt =
        instantaneousForwardDerivative(t);

    double convexity;

    if(std::abs(a) < 1e-12){

        convexity =
            sigma
            *
            sigma
            *
            t;

    } else {

        convexity =
            sigma
            *
            sigma
            /
            (2.0 * a)
            *
            (
                1.0
                -
                std::exp(
                    -2.0 * a * t
                )
            );
    }

    return
        dfdt
        +
        a * f
        +
        convexity;
}

/*
    Short-rate drift:

        theta(t) - a r(t)
*/

double HullWhiteModel::shortRateDrift(
    double t,
    double r
) const {

    return
        theta(t)
        -
        a * r;
}

/*
    Euler step:

        r_{t+dt}
        =
        r_t
        +
        [theta(t) - a r_t] dt
        +
        sigma sqrt(dt) Z
*/

double HullWhiteModel::evolveShortRate(
    double t,
    double r,
    double dt,
    double z
) const {

    return
        r
        +
        shortRateDrift(t, r) * dt
        +
        sigma
        *
        std::sqrt(dt)
        *
        z;
}
